solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -162,13 +161,78 @@ pub mod solana_ibc {
         ctx: Context<'a, 'a, 'a, 'info, Deliver<'info>>,
         message: ibc::MsgEnvelope,
     ) -> Result<()> {
+        if matches!(
+            message,
+            ibc::MsgEnvelope::Client(ibc::ClientMsg::UpdateClient(_))
+        ) {
+            panic!(
+                "Invalid message, Update Client message cannot be passed \
+                 here. Call send_update_client instead."
+            );
+        }
         let mut store = storage::from_ctx!(ctx, with accounts);
         let mut router = store.clone();
         ::ibc::core::entrypoint::dispatch(&mut store, &mut router, message)
             .map_err(error::Error::ContextError)
             .map_err(move |err| error!((&err)))
     }
 
+    /// Method to verify ed25519 signatures and send update client message.
+    ///
+    /// Since we cannot use ed25519 signature verification due to compute budget limit,
+    /// we have to call ed25519 program and pass the message, pubkey and signatures in
+    /// the same transaction in which the below method would be called. We then verify
+    /// the signatures comparing to the data in the previous instructions. If they
+    /// signature verification is success, we send the `UpdateClient` msg. This method
+    /// can only be called to send `UpdateClient` message. Calling otherwise would
+    /// panic.
+    pub fn send_update_client(
+        ctx: Context<SendUpdateClient>,
+        update_client_msg: ibc::MsgEnvelope,
+        signatures_data: Vec<SignatureData>,
```

Comment:
This needs to be read from the update_client_msg, otherwise user can pass different signatures in the message and here.  And if we do that, we actually won’t need a separate send_update_client entry point.

Reply:
ya true, u have a point. Let me try to do that then.